Dockerfile

```diff
@@ -18,13 +16,18 @@ RUN go mod download
 # Copy the go source
 COPY . .
 
+ARG TARGETARCH TARGETOS
+
 # Build
-RUN export GOOS=$(echo ${TARGETPLATFORM} | cut -d / -f1) && \
-  export GOARCH=$(echo ${TARGETPLATFORM} | cut -d / -f2) && \
-  GOARM=$(echo ${TARGETPLATFORM} | cut -d / -f3 | cut -c2-) && \
+ENV CGO_ENABLED=0 GOOS="$TARGETOS" GOARCH="$TARGETARCH"
+RUN export GOARM=$(echo ${TARGETPLATFORM} | cut -d / -f3 | cut -c2-) && \
   go build -a -o manager main.go && \
   go build -a -o github-webhook-server ./cmd/githubwebhookserver
 
+## Compress binary with upx https://github.com/upx/upx/
+RUN upx -9 /workspace/manager || true && \
+  upx -9 /workspace/github-webhook-server || true
```

Comment:
Thansks for the suggestion! But at this time I"m not yet convinced to use upx.

You said:

> Add upx to compress binaries which make final container image (extracted) 41MB from 97MB without any performance penalty. Not working correctly in arm right now but I'm working on that. more https://github.com/upx/upx/

I'm mainly concerned about the downloaded container image size(which is a set of image layers that are gzipped at the end as I understand) and I don't think compressing the executable binary results in a huge difference in the compressed, downloaded container image size.

Reply:
Well we are using `distroless/static` so the only thing we can compress is binaries right. :arrow_down: you can see UPX compression ratio for our binaries is `~39.38%` which is great right? That will directly downgrade the container's size at least `~39.38%`.

![image](https://user-images.githubusercontent.com/68642400/146889564-c91ff4d6-8a4b-46d3-991c-a9cff9e5a4ed.png)

---
I have seen people also maintaining UPX'ed binaries and container separately, so we can do that also if you want. I'm In for UPX.

Reply:
Thanks- Great. The compressed docker image size for actions-runner-controller is like 38MB these days. So upx is compressing it better than docker's native image compression!

I don't know all the side-effects of doing it though. I heard upx makes the resulting executable to decompress itself in-memory before running the original code. So would there be a bit CPU overhead and a tiny surge in disk read due to all the data in the container fs needs to be read on startup? Perhaps it can also affect some anti-virus or anti-malware detection to misbehave?

That said, it would be great if there were both upx-enabled version of the actions-runer-controller image and the standard version of it at the same time.

Reply:
> I don't know all the side-effects of doing it though. I heard upx makes the resulting executable to decompress itself in-memory before running the original code. So would there be a bit CPU overhead and a tiny surge in disk read due to all the data in the container fs needs to be read on startup? 

As long as someone doesn't run this in a very tight environment (i.e., edge)!!, we are good I guess.

> Perhaps it can also affect some anti-virus or anti-malware detection to misbehave?

Have to check in someone (I'm daily-ing Linux) else's Windows system with defender and other antivirus software,

> Thanks- Great. The compressed docker image size for actions-runner-controller is like 38MB these days. So upx is compressing it better than docker's native image compression!

Still, I haven't added go build flags `-trimpath -ldflags "-s -w"`, Testin it now

> That said, it would be great if there were both upx-enabled version of the actions-runer-controller image and the standard version of it at the same time.

I'm interested in contributing

Reply:
:scream:  with GO Build flags `-trimpath -ldflags "-s -w"` (`~35MB`) + `UPX`ed  = `~8.5MB`

![image](https://user-images.githubusercontent.com/68642400/146898140-64ddc795-88a5-4388-82fd-c3e02c17e622.png)

that makes container image(extracted) only `19.6MB` :scream: 

![image](https://user-images.githubusercontent.com/68642400/146898340-7cb02a28-55ea-4a1b-968c-9297e7c4546d.png)

> FYI, i don't know much about ``-trimpath -ldflags "-s -w"`` also :sweat_smile:  but i've seen people using it all the time

Reply:
>> Perhaps it can also affect some anti-virus or anti-malware detection to misbehave?

> Have to check in someone (I'm daily-ing Linux) else's Windows system with defender and other antivirus software,

for what it's worth sometime ago we used upx in a windows project - but our company antivirus detected from time to time the binary as virus - and unfortunately we dropped using upx for windows

Reply:
@fgalind1 What if we sign the `UPX`ed binary ? Does that help in Windows systems?